edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/IncidentTaskEx.cs

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public static class IncidentTaskEx

```

Comment:
Do we need this?

Reply:
My worry about using the existing TimeoutAfter is that it throws a timeout exception. We handle the TimeoutException in the code, and also the SDK throws TimeoutException sometimes (i believe it even converts some exception to TimeoutException because we have some logic around it?), but the point is that:

- if our original logic around the TimeoutException kicks in earlier than when we want to handle the SDK-hang problem, then that interfere with our logic (e.g. it does some retry, prolonging the time we recover)
- if the new logic that closes the endpoint kicks in sooner, then every case when the sdk throws a timeout and is handled by our original logic, that now will be caught by the new logic.

So I think it is better to use a different exception. We cannot catch the TimeoutException at the place of the SendEventAsyc call, because then we don't know that it was thrown by the sdk or by the TimeoutAfter call.

Because of this, I feel safer to use a dedicated exception for this case, even if it is ugly to duplicate the TimeoutAfter implementation. (Or maybe there can be a version of the original TimeoutAfter() that gets an action parameter, and the action parameter would throw our dedicated exception. In that case the code could go to the TaskEx.cs as it is general enough. Like 

```
public static async Task<T> TimeoutAfter<T>(this Task<T> task, Action action, TimeSpan timeout)
        {
            using (var cts = new CancellationTokenSource())
            {
                Task timerTask = Task.Delay(timeout, cts.Token);
                Task completedTask = await Task.WhenAny(task, timerTask);
                if (completedTask == timerTask)
                {
                    action();
                }

                cts.Cancel();
                return await task;
            }
        }
```

Reply:
Chatted with Viktor about this, and instead of using an action, I am passing in the error as a parameter. Same idea